### ISSUE TITLE
[Bug] Broken `model_lib` support in Python ChatConfig API

### ISSUE THREAD

**Issue by junrushao** [mlc-ai/mlc-llm#1080]
## 🐛 Bug

TL;DR. The Python ChatConfig API allows to configure `model_lib`, but not in a way expected by most of end users who just wanted to provide a relative or absolute path.

## To Reproduce

To reproduce this issue, please use the following Python script:

```python
from mlc_chat import ChatConfig, ChatModule

def main():
    """The main function that runs the benchmarking."""
    model = "./dist/mlc-chat-Llama-2-7b-chat-hf-q4f16_1"
    model_lib = "./dist/Llama-2-7b-chat-hf-q4f16_1/Llama-2-70b-chat-hf-q4f16_1-cuda.so"
    prompt = "What is the meaning of life?"
    chat_module = ChatModule(
        model=model,
        device="cuda:0",
        chat_config=ChatConfig(model_lib=model_lib),
    )
    output = chat_module.benchmark_generate(prompt, generate_length=256)

if __name__ == "__main__":
    main()
```

And the model could be obtained via:

```bash
git clone https://huggingface.co/mlc-ai/mlc-chat-Llama-2-7b-chat-hf-q4f16_1
```

and the model lib could be compiled using:

```bash
python -m mlc_llm.build --build-model-only --use-cache 0 \
    --model ./dist/models/Llama-2-7b-chat-hf/ \
    --quantization q4f16_1 \
    --max-seq-len 2048 \
    --target cuda
```

The error message is trying to give me a pretty long lecture:

```
  File "/home/junrushao/Projects/mlc-llm/python/mlc_chat/chat_module.py", line 735, in __init__
    self.lib_path = _get_lib_module_path(
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/home/junrushao/Projects/mlc-llm/python/mlc_chat/chat_module.py", line 489, in _get_lib_module_path
    raise FileNotFoundError(err_msg)
FileNotFoundError: Cannot find the model library that corresponds to `./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so`.
`./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so` is either provided in the `chat_config` you passed in, or specified in ./dist/mlc-chat-Llama-2-70b-chat-hf-q4f32_1/mlc-chat-config.json.
We searched over the following possible paths:
- ./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so-cuda.so
- dist/prebuilt/lib/./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so-cuda.so
- dist/./dist/mlc-chat-Llama-2-70b-chat-hf-q4f32_1/./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so-cuda.so
- ./dist/mlc-chat-Llama-2-70b-chat-hf-q4f32_1/./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so-cuda.so
- /opt/scratch/junrushao/mlc-llm/dist/./dist/Llama-2-70b-chat-hf-q4f32_1/Llama-2-70b-chat-hf-q4f32_1-cuda.so-cuda.so
If you would like to directly specify the model library path, you may consider passing in the `lib_path` parameter.
Please checkout https://github.com/mlc-ai/notebooks/blob/main/mlc-llm/tutorial_chat_module_getting_started.ipynb for an example on how to load a model.
```

While I understand that the tweak below should work:

```diff
-model_lib = "./dist/Llama-2-7b-chat-hf-q4f16_1/Llama-2-70b-chat-hf-q4f16_1-cuda.so"
+model_lib = "./dist/Llama-2-7b-chat-hf-q4f16_1/Llama-2-70b-chat-hf-q4f16_1" 
```

I found it quite counter-intuitive to ask me truncate a path that I believe is completely valid.

## Expected behavior

Accept a path please :)

More broadly, a question to be asked is that how “intelligent” our path finding is supposed to be? If we compile model lib locally and use pre-quantized weights, I didn’t see the point beyond asking for a full path; instead if we use prebuilt model lib, instead of cloning a whole repo, we may ask to wget a specific .so file, and in this case, we may not need intelligent path finding either.

**Comment by junrushao** [comment]
CC: @tqchen @CharlieFRuan

**Comment by junrushao** [comment]
Note that in MLC LLM, our official terminology for a shared library is “model library”, and the official customization point for a Python API is ChatConfig. This is why I, as the core developer, find `ChatConfig.model_lib` in the first place. Meanwhile, `lib_path` as the argument does exist, but seems pretty generic and could mean many things at first glance, and also, as the user, I would only focus on `ChatConfig` to properly configure the chat.

I get it that `model_lib` and `local_id` in `mlc-chat-config.json` are both used somehow in iOS/Android or other apps, but it is less self-explanatory and I have to admit that it’s necessary even for me to look up the actual code to understand what those two fields are used for…

Perhaps as a meta question, we might want to think out of the box about the design of “mlc-chat-config.json” - which fields are actually useful to the end users and which are instead metadata that could be provided in a different place.

**Comment by CharlieFRuan** [comment]
Thank you for raising the issue @junrushao, and thanks @MasterJH5574 for the input!! I agree that `model_lib` and `lib_path` can be confusing from a Python user POV.

I think the original intention of introducing `ChatConfig` was to entirely replicate `mlc-chat-config.json`, so the semantics of each field in `ChatConfig` is one-to-one to that in `mlc-chat-config.json`. As a result, as long as `lib_path` is not passed in, we use `model_lib` (either the original one in `mlc-chat-config.json`, or the one that user passed in to `ChatConfig`) to perform a search.

The search for model library path using `model_lib` in Python:
https://github.com/mlc-ai/mlc-llm/blob/3aefd9f9f25debe6f55e92b4e181cf3096f1d5e0/python/mlc_chat/chat_module.py#L388

is almost identical to that in the CLI counterpart:
https://github.com/mlc-ai/mlc-llm/blob/3aefd9f9f25debe6f55e92b4e181cf3096f1d5e0/cpp/cli_main.cc#L370

---

Now thinking about it, there probably isn't a case where Python users would want to override `ChatConfig`'s `model_lib` instead of passing in `lib_path` (so solution 2 below might be better?)

For solutions, I think there are two possibilities:
1. Include the possibility that `model_lib` is a path (absolute or relative) when searching in https://github.com/mlc-ai/mlc-llm/blob/3aefd9f9f25debe6f55e92b4e181cf3096f1d5e0/python/mlc_chat/chat_module.py#L388
2. Remove `model_lib` from `ChatConfig`, so that Python users only have two options: either specify `lib_path`, or use the `model_lib` in `mlc-chat-config.json`

Regarding whether `model_lib` should be in `mlc-chat-config.json` at all, one of the usecases (besides iOS/Android apps as @MasterJH5574 mentions, which I am not too familiar with) is that when users follow the [Getting Started](https://llm.mlc.ai/docs/) workflow, they do not need to specify a lib_path even if the `model_lib` is different from the model name (e.g. all the Llama variants) since the information is included in `mlc-chat-config.json`.

**Comment by junrushao** [comment]
Thanks @MasterJH5574 and @CharlieFRuan for getting back to me super quickly! Here is what I think:

In the short term, we are laser-focused on polishing the Python API making it perfect, and the principle I would prefer is that there is only one way to do things to avoid confusion. Based on this principle, we should:

- Remove `ChatConfig.model_lib` because users don't have to know anything about it anyways in Python API
- Rename `lib_path` to `model_lib` because it is the actual model lib we allow users to tweak with

which is basically Option 2 plus renaming.

In the long run, I believe it would make the most sense to clean up `mlc-chat-config.json`, categorize the configuration a bit into (tentatively a mind dump):
- Inherit model properties
- Configuration of the text generator
- Immutable metadata

**Comment by CharlieFRuan** [comment]
Got it, will follow up with a PR soon!

**Comment by rickzx** [comment]
Thanks @junrushao @MasterJH5574 and @CharlieFRuan for the input. I'd also like to add some input here since I'm currently working on addressing a couple of `model_lib` cannot found issue: (https://github.com/mlc-ai/mlc-llm/issues/947, https://github.com/mlc-ai/mlc-llm/issues/876, etc.)

1. Regarding `model` vs `model_lib` vs `lib_path`, we currently enforce the following convention:
- `model` is either the model name or the full path to the model directory that contains the `mlc-chat-config.json` file. In the former case, `chat_module` will auto infer the model directory. This property is directly passed into the `ChatModule` class: https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/chat_module.py#L670.
- `model_lib` is simply the model library name. It should **not** be a full path, really, according to the current logic. This is because this `model_lib` is used to infer the actual model lib file name by appending the `device_name` at the end: https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/chat_module.py#L441. And then the actual model lib file name is searched inside possible model lib paths. So this field should just be plain `Llama-2-70b-chat-hf-q4f16_1`
- `lib_path` is there in case the auto-search of `model_lib` in inferred paths doesn't work. So we have this mechanism to skip the auto-search by directly providing a full path, e.g. `./dist/Llama-2-13b-chat-hf-q4f16_1/Llama-2-13b-chat-hf-q4f16_1-cuda.so`.

2. So, it does seem that `lib_path` has its own value in addition to `model_lib` in our current logic. Simply renaming `lib_path` to `model_lib` won't work because we currently rely on `model_lib` for the auto-search logic, and `lib_path` for skipping auto-search. My suggestion is to rename `lib_path` to something more obvious, like `model_lib_override_path`. And hint the user in the error message to use this option when auto-search cannot find the model lib.

3. There is a discrepancy in C++ CLI and Python API. The C++ CLI does NOT have the functionality to override the `lib_path`. Therefore, the dist directory has to follow a certain structure. See: https://github.com/mlc-ai/mlc-llm/issues/876#issuecomment-1753322753. This is actually the reason why many CLI user encounter the issue of model lib not found. I think we do need to add a similar `lib_path` override mechanism to C++ CLI as well.

**Comment by CharlieFRuan** [comment]
Thanks @rickzx!

2. For the change, I believe we will still use the `model_lib` to do the searching, but this `model_lib` will be directly read from `mlc-chat-config.json` and we disable the option of overriding this field `ChatConfig`.  We will prioritize `lib_path` (i.e. if specified, will not perform a search with `model_lib`).
3. I agree that it would be helpful for CLI users to be able to specify a model library path, we should add that!

**Comment by junrushao** [comment]
Closed via #1086
